tests/Unit/NumericalAlgorithms/LinearSolver/Test_Gmres.cpp

```diff
@@ -0,0 +1,129 @@
+#include "Informer/Verbosity.hpp"
+#include "NumericalAlgorithms/Convergence/Criteria.hpp"
+#include "NumericalAlgorithms/Convergence/Reason.hpp"
+#include "NumericalAlgorithms/LinearSolver/Gmres.hpp"
+#include "Utilities/Gsl.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+#include "tests/Utilities/MakeWithRandomValues.hpp"
+
+namespace LinearSolver {
+namespace Serial {
+
+namespace {
+struct ScalarField : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+template <typename Tag>
+struct SomePrefix : db::PrefixTag {
+  using type = tmpl::type_from<Tag>;
+};
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.LinearSolver.Serial.Gmres",
+                  "[Unit][NumericalAlgorithms][LinearSolver]") {
+  {
+    INFO("Solve a symmetric 2x2 matrix");
+    DenseMatrix<double> matrix(2, 2);
+    matrix(0, 0) = 4.;
+    matrix(0, 1) = 1.;
+    matrix(1, 0) = 1.;
+    matrix(1, 1) = 3.;
+    const DenseVector<double> source{1., 2.};
+    const DenseVector<double> initial_guess{2., 1.};
+    const DenseVector<double> expected_solution{0.0909090909090909,
+                                                0.6363636363636364};
+    const Convergence::Criteria convergence_criteria{2, 1.e-14, 0.};
+    Gmres<DenseVector<double>> gmres{convergence_criteria,
+                                     ::Verbosity::Verbose};
+    const auto linear_operator =
+        [&matrix](const DenseVector<double>& arg) noexcept {
+          return matrix * arg;
+        };
+    auto solution = gmres(linear_operator, source, initial_guess);
+    CHECK(gmres.convergence_reason() == Convergence::Reason::AbsoluteResidual);
+    CHECK(gmres.num_iterations() == 2);
+    CHECK_ITERABLE_APPROX(solution, expected_solution);
+    {
+      INFO("Check that a solved system terminates early");
+      auto another_solution = gmres(linear_operator, source, solution);
+      CHECK(gmres.convergence_reason() ==
+            Convergence::Reason::AbsoluteResidual);
+      CHECK(gmres.num_iterations() == 0);
+      CHECK_ITERABLE_APPROX(solution, another_solution);
+    }
+  }
+  {
+    INFO("Solve a non-symmetric 2x2 matrix");
+    DenseMatrix<double> matrix(2, 2);
+    matrix(0, 0) = 4.;
+    matrix(0, 1) = 1.;
+    matrix(1, 0) = 3.;
+    matrix(1, 1) = 1.;
+    const DenseVector<double> source{1., 2.};
+    const DenseVector<double> initial_guess{2., 1.};
+    const DenseVector<double> expected_solution{-1., 5.};
+    const Convergence::Criteria convergence_criteria{2, 1.e-14, 0.};
+    Gmres<DenseVector<double>> gmres{convergence_criteria,
+                                     ::Verbosity::Verbose};
+    auto solution = gmres(
+        [&matrix](const DenseVector<double>& arg) noexcept {
+          return matrix * arg;
+        },
+        source, initial_guess);
+    CHECK(gmres.convergence_reason() == Convergence::Reason::AbsoluteResidual);
+    CHECK(gmres.num_iterations() == 2);
+    CHECK_ITERABLE_APPROX(solution, expected_solution);
+  }
+  {
+    INFO("Solve a matrix-free linear operator");
```

Comment:
Is this meaningfully different from the previous test?  The matrix isn't passed to the gmres object in either case.  The only change appears to be whether the test itself performs a matrix multiplication or writes the same multiplication out component by component.

Reply:
right, deleted this test